ansible/tasks/setup-postgres.yml

```diff
@@ -1,127 +1,35 @@
-  when: ansible_distribution_version != "18.04"
+    name: gnupg
 
-- name: Download LLVM & Clang - Ubuntu 18.04
-  apt:
-    pkg:
-      - llvm-12-dev
-      - clang-12
-  when: ansible_distribution_version == "18.04"
+- name: Add Postgres GPG key
+  apt_key:
+    keyserver: keyserver.ubuntu.com
+    id: B97B0AFCAA1A47F044F244A07FCC7D46ACCC4CF8
 
-- name: Download GCC 10
+- name: Add Postgres PPA
+  apt_repository:
+    repo: "deb https://apt-archive.postgresql.org/pub/repos/apt {{ ansible_distribution_release }}-pgdg-archive main"
+    state: present
+
+- name: Postgres - install commons
   apt:
-    pkg:
-      - gcc-10
-      - g++-10
+    name: postgresql-common
+    state: latest
+    update_cache: yes
 
-- name: Switch to GCC 10
+- name: Do not create main cluster
   shell:
-    cmd: update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-10 100 --slave /usr/bin/g++ g++ /usr/bin/g++-10 --slave /usr/bin/gcov gcov /usr/bin/gcov-10
+    cmd: sed -ri 's/#(create_main_cluster) .*$/\1 = false/' /etc/postgresql-common/createcluster.conf
 
-# Setup permissions
-- name: Update permissions for /var/tmp directory
-  file:
-    path: /var/tmp/
-    owner: root
-    group: root
-    mode: '1777'
-  become: yes
+- name: Postgres - install server
+  apt:
+    name: postgresql-{{ postgresql_major }}={{ postgresql_release }}-1.pgdg20.04+1
 
-# Building Postgres from source
-- name: Postgres - download latest release
-  get_url:
-    url: https://ftp.postgresql.org/pub/source/v{{ postgresql_release }}/postgresql-{{ postgresql_release }}.tar.gz
-    dest: /tmp
-    checksum: "{{ postgresql_release_checksum }}"
-    timeout: 60
-
-- name: Postgres - unpack archive
-  unarchive:
-    remote_src: yes
-    src: /tmp/postgresql-{{ postgresql_release }}.tar.gz
-    dest: /tmp
-
-- name: Setting CFLAGS (arm)
-  set_fact:
-    cflags: "-moutline-atomics -mtune=native -march=native -mcpu=native -fsigned-char -O2"
-  when: platform == "arm64" and ansible_distribution_version != "18.04"
-
-- name: Setting CFLAGS (arm) - Ubuntu 18.04
-  set_fact:
-    cflags: "-moutline-atomics -fsigned-char -O2"
-  when: platform == "arm64" and ansible_distribution_version == "18.04"
-
-- name: Setting CFLAGS (x86)
-  set_fact:
-    cflags: "-fsigned-char"
-  when: platform == "amd64"
-
-- name: Setting LLVM_CONFIG
-  set_fact:
-    llvm_config: "/usr/bin/llvm-config-11"
-    clang: "/usr/bin/clang-11"
-  when: ansible_distribution_version != "18.04"
-
-- name: Setting LLVM_CONFIG - Ubuntu 18.04
-  set_fact:
-    llvm_config: "/usr/bin/llvm-config-12"
-    clang: "/usr/bin/clang-12"
-  when: ansible_distribution_version == "18.04"
-
-- name: Postgres - configure
+- name: Create symlink to /usr/lib/postgresql/bin
   shell:
-    cmd: CFLAGS='{{ cflags }}' LLVM_CONFIG='{{ llvm_config }}' CLANG='{{ clang }}' ./configure --with-llvm --with-openssl --with-systemd --with-uuid=e2fs --with-libxml --with-icu --with-lz4 --exec-prefix=/usr/lib/postgresql --datarootdir=/var/lib/postgresql
```

Comment:
Yup the official docker image downloads from PPA and are additive to our flags.

Reply:
- Build configs from arm64 PPA:
```bash
$ docker run --rm -t supabase/postgres:15.1.0.84 pg_config
BINDIR = /usr/lib/postgresql/15/bin
DOCDIR = /usr/share/doc/postgresql-doc-15
HTMLDIR = /usr/share/doc/postgresql-doc-15
INCLUDEDIR = /usr/include/postgresql
PKGINCLUDEDIR = /usr/include/postgresql
INCLUDEDIR-SERVER = /usr/include/postgresql/15/server
LIBDIR = /usr/lib/aarch64-linux-gnu
PKGLIBDIR = /usr/lib/postgresql/15/lib
LOCALEDIR = /usr/share/locale
MANDIR = /usr/share/postgresql/15/man
SHAREDIR = /usr/share/postgresql/15
SYSCONFDIR = /etc/postgresql-common
PGXS = /usr/lib/postgresql/15/lib/pgxs/src/makefiles/pgxs.mk
CONFIGURE =  '--build=aarch64-linux-gnu' '--prefix=/usr' '--includedir=${prefix}/include' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--sysconfdir=/etc' '--localstatedir=/var' '--disable-silent-rules' '--libdir=${prefix}/lib/aarch64-linux-gnu' '--runstatedir=/run' '--disable-maintainer-mode' '--disable-dependency-tracking' '--with-tcl' '--with-perl' '--with-python' '--with-pam' '--with-openssl' '--with-libxml' '--with-libxslt' '--mandir=/usr/share/postgresql/15/man' '--docdir=/usr/share/doc/postgresql-doc-15' '--sysconfdir=/etc/postgresql-common' '--datarootdir=/usr/share/' '--datadir=/usr/share/postgresql/15' '--bindir=/usr/lib/postgresql/15/bin' '--libdir=/usr/lib/aarch64-linux-gnu/' '--libexecdir=/usr/lib/postgresql/' '--includedir=/usr/include/postgresql/' '--with-extra-version= (Ubuntu 15.3-1.pgdg20.04+1)' '--enable-nls' '--enable-thread-safety' '--enable-debug' '--enable-dtrace' '--disable-rpath' '--with-uuid=e2fs' '--with-gnu-ld' '--with-gssapi' '--with-ldap' '--with-pgport=5432' '--with-system-tzdata=/usr/share/zoneinfo' 'AWK=mawk' 'MKDIR_P=/bin/mkdir -p' 'PROVE=/usr/bin/prove' 'PYTHON=/usr/bin/python3' 'TAR=/bin/tar' 'XSLTPROC=xsltproc --nonet' 'CFLAGS=-g -O2 -fstack-protector-strong -Wformat -Werror=format-security' 'LDFLAGS=-Wl,-Bsymbolic-functions -Wl,-z,relro -Wl,-z,now' '--enable-tap-tests' '--with-icu' '--with-llvm' 'LLVM_CONFIG=/usr/bin/llvm-config-10' 'CLANG=/usr/bin/clang-10' '--with-lz4' '--with-zstd' '--with-systemd' '--with-selinux' 'build_alias=aarch64-linux-gnu' 'CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2' 'CXXFLAGS=-g -O2 -fstack-protector-strong -Wformat -Werror=format-security'
CC = gcc
CPPFLAGS = -Wdate-time -D_FORTIFY_SOURCE=2 -D_GNU_SOURCE -I/usr/include/libxml2
CFLAGS = -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wendif-labels -Wmissing-format-attribute -Wimplicit-fallthrough=3 -Wcast-function-type -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -moutline-atomics -g -g -O2 -fstack-protector-strong -Wformat -Werror=format-security
CFLAGS_SL = -fPIC
LDFLAGS = -Wl,-Bsymbolic-functions -Wl,-z,relro -Wl,-z,now -L/usr/lib/llvm-10/lib -Wl,--as-needed
LDFLAGS_EX = 
LDFLAGS_SL = 
LIBS = -lpgcommon -lpgport -lselinux -lzstd -llz4 -lxslt -lxml2 -lpam -lssl -lcrypto -lgssapi_krb5 -lz -lreadline -lpthread -lrt -ldl -lm
```

- Build configs from AMI:
```bash
ubuntu@db-cimpzzikygouamvsjwxa:~$ pg_config
BINDIR = /usr/lib/postgresql/bin
DOCDIR = /var/lib/postgresql/doc
HTMLDIR = /var/lib/postgresql/doc
INCLUDEDIR = /usr/local/pgsql/include
PKGINCLUDEDIR = /usr/local/pgsql/include
INCLUDEDIR-SERVER = /usr/local/pgsql/include/server
LIBDIR = /usr/lib/postgresql/lib
PKGLIBDIR = /usr/lib/postgresql/lib
LOCALEDIR = /var/lib/postgresql/locale
MANDIR = /var/lib/postgresql/man
SHAREDIR = /var/lib/postgresql
SYSCONFDIR = /usr/local/pgsql/etc
PGXS = /usr/lib/postgresql/lib/pgxs/src/makefiles/pgxs.mk
CONFIGURE =  '--with-llvm' '--with-openssl' '--with-systemd' '--with-uuid=e2fs' '--with-libxml' '--with-icu' '--with-lz4' '--exec-prefix=/usr/lib/postgresql' '--datarootdir=/var/lib/postgresql' 'CFLAGS=-moutline-atomics -mtune=native -march=native -mcpu=native -fsigned-char -O2' 'LLVM_CONFIG=/usr/bin/llvm-config-11' 'CLANG=/usr/bin/clang-11'
CC = gcc
CPPFLAGS = -D_GNU_SOURCE -I/usr/include/libxml2
CFLAGS = -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Werror=vla -Wendif-labels -Wmissing-format-attribute -Wimplicit-fallthrough=3 -Wcast-function-type -Wformat-security -fno-strict-aliasing -fwrapv -fexcess-precision=standard -Wno-format-truncation -Wno-stringop-truncation -moutline-atomics -mtune=native -march=native -mcpu=native -fsigned-char -O2
CFLAGS_SL = -fPIC
LDFLAGS = -L/usr/lib/llvm-11/lib -Wl,--as-needed -Wl,-rpath,'/usr/lib/postgresql/lib',--enable-new-dtags
LDFLAGS_EX =
LDFLAGS_SL =
LIBS = -lpgcommon -lpgport -llz4 -lxml2 -lssl -lcrypto -lz -lreadline -lpthread -lrt -ldl -lm
```

The following CFLAGS are used by our AMI build but not by official PPA
1. `-mtune=native -march=native -mcpu=native`
2. `-fsigned-char`

The `mtune`, `march`, and `mcpu` flags optimise for CPU of the **builder** machine so we _might_ see a performance boost if the AMI is run on the same AWS instance type as the builder. Given the various compute sizes offered by AWS for upgrades, its benefit really varies from project to project. Hence, I'm in favour of dropping these flags for a more generic arm64 build.

The `signed-char` flag seems to be the [default for x86 but not for arm64](https://news.ycombinator.com/item?id=18269886). I don't think it affects the integrity of the compiled binary because the official package is built in a platform specific way. Tagging @dragarcia just in case I missed anything.

Reply:
We're using Graviton3 machines, which all run of Neoverse-N1 CPUs 🤔 would be curious to run a few benchmarks just to see if there's any noticeable difference.

Will test `15.1.0.85-rc0` on staging against the currently deployed image and circle back!

Reply:
The PPA-sourced Postgres seems to have higher throughput for `tcpb-like` but looses out on `select-only` and `simple-update` 🤔

tcpb-like
```
postgres@db-nfouvluloswtodcmjimp:/home/ubuntu$ pgbench -c 100 -T 60 -b tpcb-like
pgbench (15.3 (Ubuntu 15.3-1.pgdg20.04+1))
starting vacuum...end.
transaction type: <builtin: TPC-B (sort of)>
scaling factor: 1
query mode: simple
number of clients: 100
number of threads: 1
maximum number of tries: 1
duration: 60 s
number of transactions actually processed: 55968
number of failed transactions: 0 (0.000%)
latency average = 106.504 ms
initial connection time = 614.456 ms
tps = 938.928527 (without initial connection time)

postgres@db-qphkajpasgggmoonenjr:/home/ubuntu$ pgbench -c 100 -T 60 -b tpcb-like
pgbench (15.1)
starting vacuum...end.
transaction type: <builtin: TPC-B (sort of)>
scaling factor: 1
query mode: simple
number of clients: 100
number of threads: 1
maximum number of tries: 1
duration: 60 s
number of transactions actually processed: 47349
number of failed transactions: 0 (0.000%)
latency average = 125.906 ms
initial connection time = 592.728 ms
tps = 794.244696 (without initial connection time)
```

select-only
```
postgres@db-nfouvluloswtodcmjimp:/home/ubuntu$ pgbench -c 100 -T 60 -b select-only
pgbench (15.3 (Ubuntu 15.3-1.pgdg20.04+1))
starting vacuum...end.
transaction type: <builtin: select only>
scaling factor: 1
query mode: simple
number of clients: 100
number of threads: 1
maximum number of tries: 1
duration: 60 s
number of transactions actually processed: 909330
number of failed transactions: 0 (0.000%)
latency average = 6.538 ms
initial connection time = 635.066 ms
tps = 15294.214961 (without initial connection time)

postgres@db-qphkajpasgggmoonenjr:/home/ubuntu$ pgbench -c 100 -T 60 -b select-only
pgbench (15.1)
starting vacuum...end.
transaction type: <builtin: select only>
scaling factor: 1
query mode: simple
number of clients: 100
number of threads: 1
maximum number of tries: 1
duration: 60 s
number of transactions actually processed: 986558
number of failed transactions: 0 (0.000%)
latency average = 6.031 ms
initial connection time = 595.715 ms
tps = 16581.834796 (without initial connection time)
```

simple-update
```
postgres@db-nfouvluloswtodcmjimp:/home/ubuntu$ pgbench -c 100 -T 60 -b simple-update
pgbench (15.3 (Ubuntu 15.3-1.pgdg20.04+1))
starting vacuum...end.
transaction type: <builtin: simple update>
scaling factor: 1
query mode: simple
number of clients: 100
number of threads: 1
maximum number of tries: 1
duration: 60 s
number of transactions actually processed: 343008
number of failed transactions: 0 (0.000%)
latency average = 17.345 ms
initial connection time = 616.760 ms
tps = 5765.515337 (without initial connection time)

postgres@db-qphkajpasgggmoonenjr:/home/ubuntu$ pgbench -c 100 -T 60 -b simple-update
pgbench (15.1)
starting vacuum...end.
transaction type: <builtin: simple update>
scaling factor: 1
query mode: simple
number of clients: 100
number of threads: 1
maximum number of tries: 1
duration: 60 s
number of transactions actually processed: 376560
number of failed transactions: 0 (0.000%)
latency average = 15.807 ms
initial connection time = 588.914 ms
tps = 6326.249293 (without initial connection time)
```

Differences are non-trivial in either case - might be that there are some improvements bundled with PG 15.3 which can help in a mixed workload the sorts the tcpb-like suite runs.
I'd be up to getting the postgres team onboard to reach out through the mailing list so we can contribute with a Graviton-specific patch!

Regarding this - it seems that by installing the PPA we're moving to Postgres 15.3, rather than keeping 15.1 (thus diverging from the `postgres_release` variable). 

I also see that we're pinning the version:
```
postgresql-{{ postgresql_major }}={{ postgresql_release }}-1.pgdg20.04+1
``` 

but for some reason this doesn't seem to be respected?

`apt list` confirms that 15.3 is installed. Instance ref is  `nfouvluloswtodcmjimp` on staging.
```
postgres@db-nfouvluloswtodcmjimp:/home/ubuntu$ apt list --installed | grep postgresql
postgresql-15/now 15.3-1.pgdg20.04+1 arm64 [installed,local]
postgresql-client-15/now 15.3-1.pgdg20.04+1 arm64 [installed,local]
postgresql-client-common/now 249.pgdg20.04+1 all [installed,local]
postgresql-common/now 249.pgdg20.04+1 all [installed,local]
```

Reply:
I managed to obtain similar results on m6g.xlarge for pg 15.1, which is built from the latest commit here. I used The AMI version `15.1.0.86-rc1` and ran a few times for consistency.

Average latency for tspb-like workloads is ~10% faster on pre-built than custom (112ms vs 125ms). Although simple select and update are ~10% slower for pre-built, I still think they are negligible because their base numbers are much lower, it only adds about 2ms to a simple query.

I will try to rerun the baseline with pgbench 15.3 just to make sure the client versions are also the same.

tcpb-like
```bash
postgres@db-yrxscxotxwhggepntnqk:/home/ubuntu$ pgbench -c 100 -T 60 -b tpcb-like
pgbench (15.1 (Ubuntu 15.1-1.pgdg20.04+1))
starting vacuum...end.
transaction type: <builtin: TPC-B (sort of)>
scaling factor: 1
query mode: simple
number of clients: 100
number of threads: 1
maximum number of tries: 1
duration: 60 s
number of transactions actually processed: 52901
number of failed transactions: 0 (0.000%)
latency average = 112.694 ms
initial connection time = 627.669 ms
tps = 887.357526 (without initial connection time)
```
select-only
```bash
postgres@db-yrxscxotxwhggepntnqk:/home/ubuntu$ pgbench -c 100 -T 60 -b select-only
pgbench (15.1 (Ubuntu 15.1-1.pgdg20.04+1))
starting vacuum...end.
transaction type: <builtin: select only>
scaling factor: 1
query mode: simple
number of clients: 100
number of threads: 1
maximum number of tries: 1
duration: 60 s
number of transactions actually processed: 899647
number of failed transactions: 0 (0.000%)
latency average = 6.608 ms
initial connection time = 624.868 ms
tps = 15132.534973 (without initial connection time)
```
simple-update
```bash
pgbench (15.1 (Ubuntu 15.1-1.pgdg20.04+1))
starting vacuum...end.
transaction type: <builtin: simple update>
scaling factor: 1
query mode: simple
number of clients: 100
number of threads: 1
maximum number of tries: 1
duration: 60 s
number of transactions actually processed: 333542
number of failed transactions: 0 (0.000%)
latency average = 17.834 ms
initial connection time = 628.916 ms
tps = 5607.339064 (without initial connection time)
```

Reply:
@pcnc @darora is this one good to go? I will resolve the merge conflicts one approved